presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/MergingHashAggregationBuilder.java

```diff
@@ -0,0 +1,160 @@
+ */
+package com.facebook.presto.operator.aggregation.builder;
+
+import com.facebook.presto.memory.LocalMemoryContext;
+import com.facebook.presto.operator.MergeSort;
+import com.facebook.presto.operator.OperatorContext;
+import com.facebook.presto.operator.aggregation.AccumulatorFactory;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.google.common.collect.ImmutableList;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+public class MergingHashAggregationBuilder
+    implements HashAggregationBuilder
+{
+    private final List<AccumulatorFactory> accumulatorFactories;
+    private final AggregationNode.Step step;
+    private final int expectedGroups;
+    private final ImmutableList<Integer> groupByPartialChannels;
+    private final Optional<Integer> hashChannel;
+    private final OperatorContext operatorContext;
+    private final List<Iterator<Page>> spills;
+    private InMemoryHashAggregationBuilder hashAggregationBuilder;
+    private final List<Type> groupByTypes;
+    private final LocalMemoryContext systemMemoryContext;
+    private final long maxEntriesBeforeSpill;
+    private final long memorySizeBeforeSpill;
+    private final int overwriteIntermediateChannelOffset;
+
+    public MergingHashAggregationBuilder(
+            List<AccumulatorFactory> accumulatorFactories,
+            AggregationNode.Step step,
+            int expectedGroups,
+            List<Type> groupByTypes,
+            Optional<Integer> hashChannel,
+            OperatorContext operatorContext,
+            List<Iterator<Page>> spills,
+            LocalMemoryContext systemMemoryContext,
+            long maxEntriesBeforeSpill,
+            long memorySizeBeforeSpill,
+            int overwriteIntermediateChannelOffset)
+    {
+        ImmutableList.Builder<Integer> groupByPartialChannels = ImmutableList.builder();
+        for (int i = 0; i < groupByTypes.size(); i++) {
+            groupByPartialChannels.add(i);
+        }
+
+        this.accumulatorFactories = accumulatorFactories;
+        this.step = AggregationNode.Step.partialInput(step);
+        this.expectedGroups = expectedGroups;
+        this.groupByPartialChannels = groupByPartialChannels.build();
+        this.hashChannel = hashChannel.isPresent() ? Optional.of(groupByTypes.size()) : hashChannel;
+        this.operatorContext = operatorContext;
+        this.spills = spills;
+        this.groupByTypes = groupByTypes;
+        this.systemMemoryContext = systemMemoryContext;
+        this.maxEntriesBeforeSpill = maxEntriesBeforeSpill;
+        this.memorySizeBeforeSpill = memorySizeBeforeSpill;
+        this.overwriteIntermediateChannelOffset = overwriteIntermediateChannelOffset;
+
+        rebuildHashAggregationBuilder();
+    }
+
+    @Override
+    public void processPage(Page page)
```

Comment:
why to make it implement `HashAggregationBuilder` if it doesn't support the most basic operation? Can you easily replace `MergingHashAggregationBuilder` with another `HashAggregationBuilder`? Or the other way around, can you use `MergingHashAggregationBuilder` where `HashAggregationBuilder` is normally used?